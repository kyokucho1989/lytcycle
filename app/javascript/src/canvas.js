import * as d3 from "d3";
import { routes, facilities } from "src/set_simulation_params";
import { invalidRoutesIds } from "src/consistency_check";
// データの初期値をロード

export let link, node, simulation;
import {
  setFacilityDataToModal,
  setRouteDataToModal,
  setClickEventToObject,
  facilityDialog,
  routeDialog,
} from "src/simulation";

export async function displayOperator() {
  d3.select("#svg02").append("text").attr("id", "ob1").text("作業者");
}

export async function displayStartGoalName() {
  d3.select("#svg02").selectAll("circle").select("#start").text("スタート");
  d3.select("#svg02").selectAll("circle").select("#goal").text("ゴール");
}

export async function drawLink(linksData = routes, nodesData = facilities) {
  return new Promise((resolve) => {
    d3.select("#svg02").selectAll("line").remove();
    d3.select("#svg02").selectAll("g").remove();
    function ticked() {
      link
        .attr("x1", (d) => d.source.x)
        .attr("y1", (d) => d.source.y)
        .attr("x2", (d) => adjustLinkEnd(d).x)
        .attr("y2", (d) => adjustLinkEnd(d).y);

      // node.attr("cx", (d) => d.x).attr("cy", (d) => d.y);
      node.attr("transform", (d) => `translate(${d.x}, ${d.y})`);
    }

    function adjustLinkEnd(link) {
      let x1 = link.source.x;
      let x2 = link.target.x;
      let y1 = link.source.y;
      let y2 = link.target.y;
      const offsetLength = 30;
      const distance = Math.sqrt((x2 - x1) ** 2 + (y2 - y1) ** 2);
      const ratio = (distance - offsetLength) / distance;
      const adjustedX = ratio * (x2 - x1) + x1;
      const adjustedY = ratio * (y2 - y1) + y1;
      return { x: adjustedX, y: adjustedY };
    }

    const simurateSvg = document.getElementById("svg02");
    if (simurateSvg) {
      link = d3
        .select("#svg02")
        .selectAll("line")
        .data(linksData)
        .enter()
        .append("line")
        .attr("stroke-width", (d) => (d.id.includes("re") ? 0 : 8))
        .attr("id", function (d) {
          return d.id;
        })
        .attr("marker-end", (d) => (d.id.includes("re") ? "" : "url(#arr)"));

      // node = d3
      // .select("#svg02")
      // .selectAll("circle")
      node = d3
        .select("#svg02")
        .selectAll("g") // g要素単位でデータをバインド
        .data(nodesData)
        .enter()
        .append("g") // g要素を追加して circle + text をグルーピング
        .attr("transform", (d) => `translate(${d.x}, ${d.y})`);
      // .call(
      //   d3
      //     .drag()
      //     .on("start", function () {
      //       alert("start");
      //     })
      //     .on("drag", dragged)
      //     .on("end", dragended)
      // ); // 初期位置
      // .call(
      //   d3.drag().on("drag", function (event, d) {
      //     d.x = event.x;
      //     d.y = event.y;
      //     d3.select(this).attr("transform", `translate(${d.x}, ${d.y})`);
      //   })
      // );

      // circleをg内に追加
      node
        .append("circle")
        .attr("r", (d) => d.r)
        .attr("stroke", "black") // 任意の色
        .attr("id", (d) => d.id);

      // textをg内に追加（circleの中央または外側）
      node
        .append("text")
        .text((d) => d.id)
        .attr("text-anchor", "middle") // 中央揃え
        .attr("y", 25); // 垂直方向の調整（中央に見せる）
      // .attr("font-size", "12px")
      // .attr("pointer-events", "none"); // ドラッグイベントの妨げにならないように

      // node = d3
      //   .select("#svg02")
      //   .selectAll("circle")
      //   .data(nodesData)
      //   .enter()
      //   .append("circle")
      //   .attr("r", function (d) {
      //     return d.r;
      //   })
      //   .attr("stroke", "black")
      //   .attr("id", function (d) {
      //     return d.id;
      //   });

      setNodeColor(nodesData, "#99aaee");
      setLinkColor(linksData, "#aaa");

      let invalidRoutes = linksData.filter((link) =>
        invalidRoutesIds.ids.includes(link.id)
      );
      setLinkColor(invalidRoutes, "#faa");
      // シミュレーション描画
      simulation = d3
        .forceSimulation()
        .force(
          "link",
          d3
            .forceLink()
            .strength(0)
            .id((d) => d.id)
        )
        .force("charge", d3.forceManyBody().strength(0))
        .force("x", null)
        .force("y", null)
        .on("tick", ticked);

      simulation
        .nodes(nodesData)
        .on("tick", ticked)
        .on("end", () => {
          resolve();
        });

      simulation.nodes(nodesData).on("end", () => {
        resolve();
      });
      node.call(
        d3
          .drag()
          .on("start", dragstarted)
          .on("drag", dragged)
          .on("end", dragended)
      );

      const selectMode = document.querySelector(
        'fieldset#modeSelection  input[type="radio"]:checked'
      );
      let selectModeName;

      if (selectMode !== null) {
        selectModeName = selectMode.id;
      }
      const modeState = {};
      modeState.mainState = "edit";

      switch (selectModeName) {
        case "add-facility":
          modeState.subState = "add-facility";
          break;
        case "select":
          modeState.subState = "select";
          break;
        case "add-link":
          modeState.subState = "link";
          break;
        case "delete-object":
          modeState.subState = "delete";
          break;
      }
      setClickEventToObject(modeState);
      simulation.force("link").links(linksData);

      setTimeout(() => {
        simulation.stop();
        resolve();
      }, 1000);
    } else {
      resolve();
    }
    displayStartGoalName();
    inactivePlayButtons();
  });
}

export async function setNodeColor(nodesData, color) {
  d3.selectAll("circle")
    .filter((d) => nodesData.includes(d))
    .attr("fill", color);
}

export function inactivePlayButtons() {
  const play = document.getElementById("play");
  const pause = document.getElementById("pause");
  const progress = document.getElementById("progress");
  if (play && pause && progress) {
    play.disabled = true;
    pause.disabled = true;
    progress.disabled = true;
  }
}

export function activePlayButtons() {
  const play = document.getElementById("play");
  const pause = document.getElementById("pause");
  const progress = document.getElementById("progress");
  if (play && pause && progress) {
    play.disabled = false;
    pause.disabled = false;
    progress.disabled = false;
  }
}

export async function setLinkColor(linksData, color) {
  d3.selectAll("line")
    .filter((d) => linksData.includes(d))
    .attr("stroke", color);
}

export function nodeClicked() {
  let facilityForEdit = facilities.find((facility) => facility.id == this.id);
  setFacilityDataToModal(facilityForEdit);
  facilityDialog.showModal();
}

export function linkClicked() {
  const canvasController = document.querySelector(
    "[data-controller='simulationmode']"
  ).controller;
  console.log(canvasController);

  let routeForEdit = routes.find((route) => route.id == this.id);
  setRouteDataToModal(routeForEdit);
  routeDialog.showModal();
}

function dragstarted(event) {
  if (!event.active) simulation.alphaTarget(0.3).restart();
  event.subject.fx = event.subject.x;
  event.subject.fy = event.subject.y;
}

function dragged(event) {
  event.subject.fx = event.x;
  event.subject.fy = event.y;
}

function dragended(event) {
  if (!event.active) simulation.alphaTarget(0);
  event.subject.fx = null;
  event.subject.fy = null;
}

export function changeInactiveObject() {
  d3.select("#svg02").selectAll("line").on("click", null);
  d3.select("#svg02").selectAll("circle").on("click", null);
}

export function changeActiveObject() {
  d3.select("#svg02").selectAll("line").on("click", linkClicked);
  d3.select("#svg02").selectAll("circle").on("click", nodeClicked);
}
