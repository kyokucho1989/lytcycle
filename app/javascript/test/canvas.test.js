// query utilities:
import {
  // Tip: all queries are also exposed on an object
  // called "queries" which you could import here as well
  waitFor,
} from "@testing-library/dom";
// adds special assertions like toHaveTextContent
import "@testing-library/jest-dom";

import { drawLink } from "../src/canvas.js";
let facilitiesInitial, routesInitial;
beforeEach(() => {
  facilitiesInitial = [
    {
      index: 0,
      x: 230,
      y: 310,
      r: 10,
      id: "start",
      processingTime: 1,
      type: "start",
      name: "startPoint",
    },
    {
      index: 1,
      x: 330,
      y: 60,
      r: 15,
      id: 1,
      processingTime: 15,
      type: "machine",
      name: "machine-no-1",
      isProcessing: false,
      hasMaterial: false,
      storageSize: 0,
      processingEndTime: 0,
    },
    {
      index: 2,
      x: 430,
      y: 310,
      r: 10,
      id: "goal",
      processingTime: 1,
      type: "goal",
      name: "goalPoint",
    },
  ];

  routesInitial = [
    { source: "start", target: 1, l: 20, id: "root10" },
    { source: 1, target: "goal", l: 20, id: "root11" },
    { source: "goal", target: "start", l: 20, id: "root12" },
    { source: "start", target: "goal", l: 20, id: "re_root10" },
    { source: 1, target: "start", l: 20, id: "re_root11" },
    { source: "goal", target: 1, l: 20, id: "re_root12" },
  ];
});

test("draw link and circle", async () => {
  const div = document.createElement("div");
  div.innerHTML = `
    <svg id="svg02" xmlns="http://www.w3.org/2000/svg" width="700" height="800">
    </svg>
  `;
  document.body.appendChild(div); // これを追加して、Jest の仮想 DOM に反映

  await drawLink(routesInitial, facilitiesInitial);

  await waitFor(() => {
    const group = document.querySelector("g");
    expect(group).toHaveAttribute("transform");
  });

  expect(div).toMatchSnapshot();
});
