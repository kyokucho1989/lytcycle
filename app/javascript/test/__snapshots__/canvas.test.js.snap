// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`draw link and circle 1`] = `
<div>
  
    
  <svg
    height="800"
    id="svg02"
    width="700"
    xmlns="http://www.w3.org/2000/svg"
  >
    
    
    <line
      id="root10"
      marker-end="url(#arr)"
      stroke="#aaa"
      stroke-width="8"
      x1="230"
      x2="318.8582797093769"
      y1="310"
      y2="87.8543007265578"
    />
    <line
      id="root11"
      marker-end="url(#arr)"
      stroke="#aaa"
      stroke-width="8"
      x1="330"
      x2="418.8582797093769"
      y1="60"
      y2="282.14569927344223"
    />
    <line
      id="root12"
      marker-end="url(#arr)"
      stroke="#aaa"
      stroke-width="8"
      x1="430"
      x2="260"
      y1="310"
      y2="310"
    />
    <line
      id="re_root10"
      marker-end=""
      stroke="#aaa"
      stroke-width="0"
      x1="230"
      x2="400"
      y1="310"
      y2="310"
    />
    <line
      id="re_root11"
      marker-end=""
      stroke="#aaa"
      stroke-width="0"
      x1="330"
      x2="241.1417202906231"
      y1="60"
      y2="282.14569927344223"
    />
    <line
      id="re_root12"
      marker-end=""
      stroke="#aaa"
      stroke-width="0"
      x1="430"
      x2="341.1417202906231"
      y1="310"
      y2="87.8543007265578"
    />
    <circle
      cx="230"
      cy="310"
      fill="#99aaee"
      id="start"
      r="10"
      stroke="black"
    />
    <circle
      cx="330"
      cy="60"
      fill="#99aaee"
      id="1"
      r="15"
      stroke="black"
    />
    <circle
      cx="430"
      cy="310"
      fill="#99aaee"
      id="goal"
      r="10"
      stroke="black"
    />
  </svg>
  
  
</div>
`;
